### PR TITLE
fix: Use JSONSchema to verify the original data submitted by users

### LINKED ISSUE #918: bug: config route with non-existent fields should return 400
# Bug report

## Describe the bug

config route with non-existent fields should return 400

## How to Reproduce

```
{
	caseDesc: "config route with non-existent fields (r1)",
	Object:   MangerApiExpect(t),
	Path:     "/apisix/admin/routes/r1",
	Method:   http.MethodPut,
	Body: `{
			"uri": "/hello",
			"nonexistent": "test non-existent",
			"upstream": {
				"type": "roundrobin",
				"nodes": [{
					"host": "172.16.238.20",
					"port": 1980,
					"weight": 1
				}]
			}
		}`,
	Headers:      map[string]string{"Authorization": token},
	ExpectStatus: http.StatusBadRequest,
},
```
it return 200 and insert it to etcd now.

### LINKED ISSUE #955: bug: JSON schema verification should be performed on the original data submitted by the user
# Bug report

## Describe the bug

JSON schema verification should be performed on the original data submitted by the user, not the data after binding structures

The usual practice in Go is to bind the request body to a structure for subsequent processing of the request body. 

This will cause a problem, that is, if there is no corresponding value in the request body for the fields defined in the structure, 
the bound object will automatically be assigned a value of zero. 

At this time, there may be problems with json schema verification.

### PR DESCRIPTION
**Pull request by nic-chen**
Please answer these questions before submitting a pull request

- Why submit this pull request?
- [x] Bugfix
- [ ] New feature provided
- [ ] Improve performance

- Related issues
close #918 
close #955 
___
### Bugfix
- Description

The usual practice in Go is to bind the request body to a structure for subsequent processing of the request body.

This will cause a problem, that is, if there is no corresponding value in the request body for the fields defined in the structure,
the bound object will automatically be assigned a value of zero.

At this time, there may be problems with json schema verification.

- How to fix?

Use JSONSchema to verify the original data submitted by users

### CODE REVIEW

File: api/test/e2e/json_schema_validate_test.go [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+                                        "type": "roundrobin",
+                                        "nodes": [{
+                                                "host": "172.16.238.20",
+                                                "port": 1980,
+                                                "weight": 1
+                                        }]
+                                }
+                        }`,
+			Headers:      map[string]string{"Authorization": token},
+			ExpectStatus: http.StatusBadRequest,
```
Comment (membphis) [inline api/test/e2e/json_schema_validate_test.go:44]: need to check the response body

**Review by membphis — CHANGES_REQUESTED** [review]

File: api/filter/schema.go [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+	"upstreams": "upstream",
+	"services":  "service",
+	"consumers": "consumer",
+	"ssl":       "ssl",
+}
+
+func SchemaCheck() gin.HandlerFunc {
+	return func(c *gin.Context) {
+		pathPrefix := "/apisix/admin/"
+		resource := strings.Replace(c.Request.URL.Path, pathPrefix, "", 1)
```
Comment (tokers) [inline api/filter/schema.go:45]: Please use `strings.TrimPrefix`.

File: api/filter/schema.go [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+		}
+
+		// other filter need it
+		c.Request.Body = ioutil.NopCloser(bytes.NewBuffer(reqBody))
+
+		validator, err := store.NewAPISIXSchemaValidator("main." + schemaKey)
+		if err != nil {
+			errMsg := fmt.Sprintf("init validator failed: %s", err)
+			c.AbortWithStatusJSON(http.StatusBadRequest, consts.InvalidParam(errMsg))
+			log.Errorf(errMsg)
```
Comment (tokers) [inline api/filter/schema.go:77]: Use `log.Error`.

Doesn't the golangci-lint find this error?

File: api/internal/core/store/validate_test.go [hunk 4]
```diff
@@ -437,3 +437,44 @@ func TestAPISIXJsonSchemaValidator_Route_checkRemoteAddr(t *testing.T) {
...
+		},
+		"desc": "test description"
+	}`
+	err = validator.Validate([]byte(reqBody))
+	assert.Nil(t, err)
+
+	// config with not exists field, should be failed.
+	reqBody = `{
+	      "username": "jack",
+              "not-exist": "val",
```
Comment (tokers) [inline api/internal/core/store/validate_test.go:465]: Bad indent.

File: api/internal/core/store/validate_test.go [hunk 5]
```diff
@@ -437,3 +437,44 @@ func TestAPISIXJsonSchemaValidator_Route_checkRemoteAddr(t *testing.T) {
...
+		      "rejected_code": 503,
+		      "key": "remote_addr"
+		  }
+		},
+		"desc": "test description"
+	}`
+	err = validator.Validate([]byte(reqBody))
+	assert.Nil(t, err)
+
+	// config with not exists field, should be failed.
```
Comment (tokers) [inline api/internal/core/store/validate_test.go:462]: not exists => non existent

**Comment by membphis** [comment]
@nic-chen I think we may need merge the `master` branch

File: api/test/e2e/json_schema_validate_test.go (hunk 1 above)
Comment (membphis) [reply api/test/e2e/json_schema_validate_test.go:44]: any news?

File: api/filter/schema.go (hunk 3 above)
Comment (nic-chen) [reply api/filter/schema.go]: fixed.

File: api/filter/schema.go (hunk 2 above)
Comment (nic-chen) [reply api/filter/schema.go]: fixed.

File: api/internal/core/store/validate_test.go (hunk 4 above)
Comment (nic-chen) [reply api/internal/core/store/validate_test.go]: fixed.

File: api/internal/core/store/validate_test.go (hunk 5 above)
Comment (nic-chen) [reply api/internal/core/store/validate_test.go]: fixed.

File: api/test/e2e/json_schema_validate_test.go (hunk 1 above)
Comment (nic-chen) [reply api/test/e2e/json_schema_validate_test.go:44]: fixed.

**Review by membphis — APPROVED** [review]

File: api/filter/schema.go [hunk 6]
```diff
@@ -0,0 +1,195 @@
...
+	"routes":    "route",
+	"upstreams": "upstream",
+	"services":  "service",
+	"consumers": "consumer",
+	"ssl":       "ssl",
+}
+
+func parseCert(crt, key string) ([]string, error) {
+	if crt == "" || key == "" {
+		return nil, errors.New("invalid certificate")
```
Comment (tokers) [inline api/filter/schema.go:48]: the error message is not accurate, should be "empty certificate or private key"

**Comment by membphis** [comment]
@tokers @ShiningRush @gxthrj please take a look at this PR, need your confirm ^_^

File: api/filter/schema.go [hunk 7]
```diff
@@ -0,0 +1,195 @@
...
+	}
+
+	x509Cert, err := x509.ParseCertificate(certDERBlock.Bytes)
+
+	if err != nil {
+		return nil, errors.New("Certificate resolution failed")
+	}
+
+	//domain
+	snis := []string{}
```
Comment (tokers) [inline api/filter/schema.go:68]: Just use `var snis []string` is OK.

File: api/filter/schema.go [hunk 8]
```diff
@@ -0,0 +1,195 @@
...
+	if x509Cert.DNSNames != nil && len(x509Cert.DNSNames) > 0 {
+		snis = x509Cert.DNSNames
+	} else if x509Cert.IPAddresses != nil && len(x509Cert.IPAddresses) > 0 {
+		for _, ip := range x509Cert.IPAddresses {
+			snis = append(snis, ip.String())
+		}
+	} else {
+		if x509Cert.Subject.Names != nil && len(x509Cert.Subject.Names) > 1 {
+			var attributeTypeNames = map[string]string{
```
Comment (tokers) [inline api/filter/schema.go:77]: use short assign way:

```
attributesTypeNames := ...
```

Comment (tokers) [reply api/filter/schema.go:77]: Why the length should be larger than 1.

File: api/internal/utils/consts/api_error.go [hunk 9]
```diff
@@ -49,7 +49,7 @@ func (err ApiError) Error() string {
 }
 
 func InvalidParam(message string) *ApiError {
-	return &ApiError{400, 400, message}
+	return &ApiError{400, 10000, message}
```
Comment (tokers) [inline api/internal/utils/consts/api_error.go:52]: Why changes the number, the number is too magic, can i reference some docs?

File: api/filter/schema.go (hunk 8 above)
Comment (nic-chen) [reply api/filter/schema.go:77]: > Why the length should be larger than 1.

fixed.

**Comment by membphis** [comment]
@nic-chen @tokers 

We can fix some minor problems later.
We can create a new issue to track these minor issues and ensure that they will not be forgotten.

**Comment by nic-chen** [comment]
> @nic-chen @tokers
> 
> We can fix some minor problems later.
> We can create a new issue to track these minor issues and ensure that they will not be forgotten.

have create a issue for it: https://github.com/apache/apisix-dashboard/issues/1023

**Comment by tokers** [comment]
> > @nic-chen @tokers
> > We can fix some minor problems later.
> > We can create a new issue to track these minor issues and ensure that they will not be forgotten.
> 
> have create a issue for it: #1023

OK

**Review by tokers — APPROVED** [review]

**Pull request merged**